[SPARK-36123][SQL] Parquet vectorized reader doesn't skip null values correctly

sunchao opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'core/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->

Fix the skipping values logic in Parquet vectorized reader when column index is effective, by considering nulls and only call `ParquetVectorUpdater.skipValues` when the values are non-null.

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->

Currently, the Parquet vectorized reader may not work correctly if column index filtering is effective, and the data page contains null values. For instance, let's say we have two columns `c1: BIGINT` and `c2: STRING`, and the following pages:
```
   * c1        500       500       500       500
   *  |---------|---------|---------|---------|
   *  |-------|-----|-----|---|---|---|---|---|
   * c2     400   300   300 200 200 200 200 200
```

and suppose we have a query like the following:
```sql
SELECT * FROM t WHERE c1 = 500
```

this will create a Parquet row range `[500, 1000)` which, when applied to `c2`, will require us to skip all the rows in `[400,500)`. However the current logic for skipping rows is via `updater.skipValues(n, valueReader)` which is incorrect since this skips the next `n` non-null values. In the case when nulls are present, this will not work correctly.

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->

No

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->

Added a new test in `ParquetColumnIndexSuite`.

sunchao committed:
initial commit

sunchao commented:
cc @dongjoon-hyun @viirya @cloud-fan

SparkQA commented:
**[Test build #140986 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/140986/testReport)** for PR 33330 at commit [`41a7ca8`](https://github.com/apache/spark/commit/41a7ca8b7b2464a7363ba6d89dfdbeb8ed7c96aa).

dongjoon-hyun commented:
Thank you for pinging me, @sunchao .

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/45500/

dongjoon-hyun reviewed (dismissed):
Hi, @sunchao . Could you make an isolated UT first before refactoring all test suites? And, please use a test prefix for the bug test case, `SPARK-36123: `.

And, cc @gengliangwang since this is created as a release blocker.

dongjoon-hyun commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetColumnIndexSuite.scala:
```diff
@@ -31,24 +31,25 @@ class ParquetColumnIndexSuite extends QueryTest with ParquetTest with SharedSpar
+      df.coalesce(1)
           .write
           .option("parquet.page.size", "4096")
+          .option("parquet.enable.dictionary", "false")
```
In terms of the test coverage, we had better test both cases, `true` and `false`.

sunchao replied:
good point - will do.

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/45500/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/45500/

sunchao commented:
thanks @dongjoon-hyun - what do you think if I open a separate PR to do the refactoring on the test suite first? it will make the changes easier.

SparkQA commented:
**[Test build #140986 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/140986/testReport)** for PR 33330 at commit [`41a7ca8`](https://github.com/apache/spark/commit/41a7ca8b7b2464a7363ba6d89dfdbeb8ed7c96aa).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

gengliangwang commented:
@sunchao Thanks for the work. I think it's OK to have a PR for test refactoring.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/140986/

sunchao commented:
thanks @gengliangwang - I opened #33334 for this.

dongjoon-hyun commented:
Thank you, @sunchao .

viirya commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetColumnIndexSuite.scala:
```diff
@@ -31,24 +31,25 @@ class ParquetColumnIndexSuite extends QueryTest with ParquetTest with SharedSpar
    *  |-------|-----|-----|---|---|---|---|---|
    * col_2   400   300   200 200 200 200 200 200
```
BTW, about the layout, the total of col_1 or col_2 is all 2000? But looks like col_2 has only 1900?

sunchao replied:
opps you are right - it should be 400, 300, 300, 200, 200, 200, 200, 200

viirya replied:
And also the PR description. :)

viirya commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedRleValuesReader.java:
```diff
@@ -255,6 +253,36 @@ private void readBatchInternal(
+      int num = Math.min(n, this.currentCount);
+      switch (mode) {
+        case RLE:
+          if (currentValue == state.maxDefinitionLevel) {
+            updater.skipValues(num, valuesReader);
+          }
```
Is else case for null value?

sunchao replied:
yes the else case is when the value is null - let me add some comments.

viirya commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedRleValuesReader.java:
```diff
@@ -255,6 +253,36 @@ private void readBatchInternal(
+          for (int i = 0; i < num; ++i) {
+            if (currentBuffer[currentBufferIdx++] == state.maxDefinitionLevel) {
+              updater.skipValues(1, valuesReader);
+            }
```
ditto.

If so, maybe it is good to add a comment.

viirya commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedRleValuesReader.java:
```diff
@@ -358,7 +386,14 @@ public void skipIntegers(int total) {
       if (this.currentCount == 0) this.readNextGroup();
       int n = Math.min(left, this.currentCount);
-      advance(n);
+      switch (mode) {
+        case RLE:
+          break;
+        case PACKED:
+          currentBufferIdx += n;
+          break;
+      }
+      currentCount -= n;
```
Hmm, this is a old bug?

sunchao replied:
no - it used to call `advance` which does the same thing, but since we removed the method I just replaced it with the old method body

viirya replied:
Do you remove `advance()` method? I don't see it is removed here.

sunchao replied:
hmm somehow it survived - let me remove it for real.

dongjoon-hyun commented:
I reviewed and merged https://github.com/apache/spark/pull/33334 .
Please rebase this PR to the master.

dongjoon-hyun commented:
Thank you for updating.

sunchao commented:
Thanks @dongjoon-hyun and @viirya ! I've rebased the PR and addressed the comments.  Please take another look.

With the current test suite it's pretty hard to add more "fine-grained" test cases. I'm working on a different test suite for the vectorized reader which provides much better coverage, which is how I uncovered this bug. I plan to post another PR for that later.

viirya commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedRleValuesReader.java:
```diff
@@ -255,6 +253,39 @@ private void readBatchInternal(
+          break;
+        case PACKED:
+          for (int i = 0; i < num; ++i) {
+            // save as above, only skip non-null values from `valuesReader`
```
save? same?

sunchao replied:
oops

sunchao committed:
fix typo

SparkQA commented:
**[Test build #141004 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141004/testReport)** for PR 33330 at commit [`cfd73a7`](https://github.com/apache/spark/commit/cfd73a73ec60eb8b44823baf60f422377e9bced9).

gengliangwang commented:
@sunchao please update the PR description for the test refactoring part.

sunchao commented:
@gengliangwang done

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/45518/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/45518/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/45518/

cloud-fan commented:
thanks, merging to master/3.2!

SparkQA commented:
**[Test build #141004 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141004/testReport)** for PR 33330 at commit [`cfd73a7`](https://github.com/apache/spark/commit/cfd73a73ec60eb8b44823baf60f422377e9bced9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/141004/

dongjoon-hyun commented:
Thank you, @sunchao and all!
